### PR TITLE
Feature/lightning

### PR DESCRIPTION
**Pull request by karpathy**
I'm trying @williamFalcon , but I have somewhat mixed feelings about it. The API are now matched up and I can train the basic loop with either:

```bash
$ USE_LIGHTNING=0 python bench.py
5 epochs took 33.225314s, or 6.645063s/epoch
```

or

```bash
$ USE_LIGHTNING=1 python bench.py
5 epochs took 30.068728s, or 6.013746s/epoch
```

some overhead incurred, not that it matters too much at the stage of a single GPU.

To merge would still have to:

- clean up a bit further and make even more transparent ideally, needs a bit more thought
- delete `bench.py`
- uprev all notebooks
- uprev Readme file

### CODE REVIEW

**Comment by karpathy** [comment]
ok with the last commit I feel a bit better about things. `mingpt/fake_lightning.py` is now basically a minLightning :D, which could ideally be imported instead of Lightning and get a strict subset of just the very basic functionality. It's not functionally equivalent and a little bit hardcoded to minGPT purposes, but that's okay.

**Comment by williamFalcon** [comment]
nice! yeah, when you enable tensorboard and checkpoints the training slows down a bit. 

we actually have tests to ensure we don’t incur a meaningful overhead over a vanilla pytorch script. lightning only adds 300ms per epoch. 

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/benchmarks/test_parity.py

**Comment by williamFalcon** [comment]
next you should convert the datasets to datamodules. this means you can 100% decouple the model from the data :)

**Comment by karpathy** [comment]
@williamFalcon  ok done. few things I find a bit gross:

- feels like I have to manually create a spurious `train_dataset`, because I need to know the size of vocab to calculate the learning rate decay, and to sample at the end etc.
- see `data_module.setup('fit') # fit... should be 'train'? ;\` . Based on this https://pytorch-lightning.readthedocs.io/en/latest/new-project.html is `'fit'` the currently sanctioned name for the training stage? Should it maybe be 'train'?

**Comment by karpathy** [comment]
Alright calling it here for today, I'm tired and still have some actual work to do. I'm pretty sure I don't understand how `pl.LightningDataModule` API is supposed to be used properly between init, prepare, setup, and dataloader calls, what stage=None means, and how it is expected to be called from the Trainer, etc. To be continued...

**Comment by williamFalcon** [comment]
ah i see the confusion.

.fit() calls setup 
and .test() also calls setup.

If you're training on 1 gpu, none of what i'm about to say matters. this only matters when making the code agnostic to n gpus 
or n tpus.

### prepare data
This is to do something only once (ie: in 100 GPU world, only on GPU 0. examples are: tokenize, download, etc...).

### setup
This is another prep stage but it's called on every GPU. This means that splitting or anything like that can be done here.

### train, val, test dataloaders
These are **lazy** called... which means that you don't have the overhead of creating the data until you really need it (this is key for performance applications).

### data for init
there's a case where your model might depend on information about the data (ie: voca size, num_classes, etc). In this case, you can just hardcode this into the datamodule:

```python
def __init__(...):
    self.num_classes = x
    self.vocab_size = y
```

or get the info in setup:

```python
def setup(...):
    download()
    tokenize()
    self.vocab = count_vocab()
```

Normally, lightning calls prepare_data and setup for you automatically in training. However, depending on how you set it up (let's say you got the vocab size in setup), then you can manually call it after init.

```python
dm = Datamodule()

# even if called manually, lightning makes sure it only happens on the correct devices
dm.prepare_data()
dm.setup()

model = LitModel(vocab_size=dm.vocab_size)
trainer = Trainer()
trainer.fit(model, dm)
```

I also don't love that you have to call prepare_data + setup yourself. Open to any ideas that you might have :) 
This is a pattern we came up with in close collaboration with our partners at Facebook.

### Datamodule examples
[Simclr made agnostic of dataset](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/pl_bolts/models/self_supervised/simclr/simclr_module.py). (it's pretty cool you can use the same code and train on any dataset without changing your original base code.    

[Imagenet datamodule](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/cf3830787e1352ce71d673347fb5648c90497b48/pl_bolts/datamodules/imagenet_datamodule.py#L11)    

[STL-10 datamodule](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/cf3830787e1352ce71d673347fb5648c90497b48/pl_bolts/datamodules/stl10_datamodule.py#L12)

### datamodule videos
and of course, the obligatory [tutorial on datamodules](https://www.youtube.com/watch?v=L---MBeSXFw) by one of our team members.

### real-world tutorials
Here's one of our new tutorials as well on implementing SimCLR which would be a more realistic complex example.
https://www.youtube.com/watch?v=p8QFB1CiAoQ

File: bench.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+        self.train_dataset = CharDataset('text8', self.block_size, 'train')
+
+    def prepare_data(self): # called only on 1 GPU/machine
+        pass # could technically download text8 here...
+
+    def setup(self, stage): # called for every GPU/machine
+        if stage == 'train' or stage == 'fit':
+            pass # nothing to do, the train_dataset is initialized in the constructor
+        elif stage == 'val':
```
Comment (williamFalcon) [inline bench.py:98]: stage here is whether the setup is happening on fit or test...

BUT... to your point, do you think it might make more sense to change the stage to 'train', 'val', 'test'? the thing is that train, val are usually handled together (ie: split train into train/val and have a separate test set)

File: bench.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        loader = DataLoader(self.test_dataset, batch_size=self.batch_size,
+                            shuffle=False, pin_memory=bool(self.pin_memory),
+                            num_workers=self.num_workers)
+        return loader
+
+
+# -----------------------------------------------------------------------------
+
+parser = argparse.ArgumentParser()
+parser.add_argument('-x', '--num-epochs', type=int, default=5, help="number of epochs to train for")
```
Comment (williamFalcon) [inline bench.py:127]: Datamodules can do:

```python
parser = argparse.ArgumentParser()

# enables whatever you have in your init in argparse :) 
parser = CharDataModule.add_argparse_args(parser)

# enable all the trainer flags in argparse
parser = Trainer.add_argparse_args(parser)

args = parser.parse_args()

# now you can init whatever objects automatically as well:
trainer = Trainer.from_argparse_args(args, any_flag_to_override=...)

dm = CharDataModule.from_argparse_args(args)
```

Which lets you do things like:

```bash
python main.py --gpus 2 --num_nodes 3 --batch_size 32

File: bench.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+            self.val_dataset = CharDataset('text8', self.block_size, 'val', override_vocab=self.train_dataset.vocab)
+        elif stage == 'test':
+            self.test_dataset = CharDataset('text8', self.block_size, 'test', override_vocab=self.train_dataset.vocab)
+        else:
+            raise ValueError(f"stage {stage} is not recognized")
+
+    def train_dataloader(self):
+        loader = DataLoader(self.train_dataset, batch_size=self.batch_size,
+                            shuffle=True, pin_memory=bool(self.pin_memory),
+                            num_workers=self.num_workers)
```
Comment (williamFalcon) [inline bench.py:108]: btw... with multiple GPUs num_workers > 0 in ddp_spawn mode is slow. This is a pytorch limitation because ddp_spawn generates subprocesses and in each subprocess there are more subprocesses generated by dataloaders.

that's why we recommend ddp as the backend for multi-gpu but unfortunately can't be called on a jupyter lab because those have limitations as well haha.

basically, until we re-invent jupyter notebooks we are a bit stuck haha...

File: mingpt/fake_lightning.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        val_loader = data_module.val_dataloader()
+        data_module.setup('test')
+        test_loader = data_module.test_dataloader()
+
+        # ship model to gpu if possible
+        device = 'cpu'
+        if self.gpus > 0 and torch.cuda.is_available():
+            logger.info("found CUDA device, shipping model to GPU")
+            device = 'cuda'
+            self.model = self.model.to(device)
```
Comment (williamFalcon) [inline mingpt/fake_lightning.py:113]: if you are enabling gpus=x i would just do ```model.cuda(x)``` so people can place models on a gpu indexed by the PCI_BUS_ID (you might need this flag enabled though

```
export CUDA_DEVICE_ORDER=PCI_BUS_ID
```

Comment (williamFalcon) [reply mingpt/fake_lightning.py:113]: but for teaching purposes it may be overkill

**Comment by williamFalcon** [comment]
> Alright calling it here for today, I'm tired and still have some actual work to do. I'm pretty sure I don't understand how `pl.LightningDataModule` API is supposed to be used properly between init, prepare, setup, and dataloader calls, what stage=None means, and how it is expected to be called from the Trainer, etc. To be continued...

yeah, this is an optional abstraction (you can use any dataloader with lightning). we also just introduced it, so need to make the docs more clear. but it really is optional, so it’s not a big deal to use dataloaders directly. it just makes the data more reusable

File: bench.py (hunk 1 above)
Comment (karpathy) [reply bench.py]: Sorry I'm pretty sure I totally misunderstood what a "stage" is and thought it referred to splits.

File: bench.py (hunk 2 above)
Comment (karpathy) [reply bench.py:89]: Neat! I'll have to read more of the docs

File: bench.py (hunk 3 above)
Comment (karpathy) [reply bench.py]: Yes, subtle point!

File: mingpt/fake_lightning.py (hunk 4 above)
Comment (karpathy) [reply mingpt/fake_lightning.py]: Thank you, yes it's cleaner. I like the more general .to syntax because we'll have many different XPUs etc, feels a bit more future proof

**Comment by karpathy** [comment]
Okay I merged one more big refactor. Honestly I am starting to think this branch was a very bad idea. I thought I could make things clean but there is a lot of baggage that Lightning "leaks" in a number of places, e.g. w.r.t. model checkpointing, the use of Training/Eval Result structures, forcing me into relatively odd looking abstractions and half-measures.

Anyway, thank you for your help @williamFalcon , I'll have to sleep on this a few days, read the Lightning docs more, and then maybe give it another shot some other time.

**Comment by williamFalcon** [comment]
Ok, so i think the original PR i submitted kept it as minimal as possible since you can keep lightning light or go deep.
The only "required" changes is the LightningModule. 

The results and datamodule stuff is optional and there more for the power users. So, if you want to keep it very minimal, i would skip the datamodule stuff and use plain dataloaders.

I added the callback just to keep the code clean, but you can also do most of it in a hook within the lightningmodule itself

```
def on_train_batch_end(self,...):
    # no callback needed, you can keep it in the lightningmodule as well
```

the result is also optional (but makes life easier), you can also just return a dictionary:
```
def training_step(...):
    return {'loss': loss}
```

File: bench.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+                     precision=args.precision, default_root_dir=args.default_root_dir)
+trainer.fit(model, train_dataloader, val_dataloader)
+t1 = time.time()
+logging.info("%d epochs took %fs, or %fs/epoch", args.num_epochs, t1 - t0, (t1-t0)/args.num_epochs)
+
+# todo below: I don't yet understand the Lightning checkpoint schema
+# logging.info("testing...")
+# ckpt_path = os.path.join(args.default_root_dir, 'model.pt')
+# model.load_from_checkpoint(ckpt_path) # load the best checkpoint we found
+# trainer.test(test_dataloader=test_dataloader)
```
Comment (williamFalcon) [inline bench.py:131]:
```suggestion
# ckpt_path = os.path.join(args.default_root_dir, 'model.pt')
# model.load_from_checkpoint(ckpt_path) # load the best checkpoint we found
# trainer.test(test_dataloader=test_dataloader)
```

```
# Note: LIGHTNING automatically loads the best checkpoint when you call .test()
trainer.test(test_dataloader=test_dataloader)
```

File: mingpt/fake_lightning.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+class EvalResult(Result):
+    pass
+
+class LightningModule(nn.Module):
+
+    def load_from_checkpoint(self, checkpoint_path):
+        logger.info("loading the best model checkpoint from %s", checkpoint_path)
+        state_dict = torch.load(checkpoint_path)
+        self.load_state_dict(state_dict)
```
Comment (williamFalcon) [inline mingpt/fake_lightning.py:40]: In lightning the checkpoint is not just the weights.

The checkpoint also has:
- all the training state (epoch, state, etc)
- hyperparams used in training
- the lightning version
- states of 16-bit scaling, etc
- states of all the learning rate schedulers, etc...

Load from checkpoint restores all that. 

If you want just the state dict then you can do:
```
ckpt = torch.load(path)
model.load_state_dict(ckpt['state_dict'])
```

File: mingpt/model.py [hunk 7]
```diff
@@ -189,9 +204,31 @@ def forward(self, idx, targets=None):
...
+
+        if split == 'train':
+            result = pl.TrainResult(minimize=loss)
+        elif split == 'val':
+            result = pl.EvalResult(checkpoint_on=loss)
+        elif split == 'test':
+            result = pl.EvalResult()
+        result.log(f'{split}_loss', loss)
+
+        return result
```
Comment (williamFalcon) [inline mingpt/model.py:225]: if you are not logging, then you can just directly return the loss. no need for a resultsdict

File: mingpt/fake_lightning.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+class Callback:
+    pass
```
Comment (williamFalcon) [inline mingpt/fake_lightning.py:43]: sorry! this is 100% optional. This is a new addition and I see we forgot to include the simple case and doc examples using a dict or the loss directly

Comment (karpathy) [reply mingpt/fake_lightning.py]: got it, ok converted to use of dicts with latest commit

File: bench.py (hunk 5 above)
Comment (karpathy) [reply bench.py:131]: got it. it looks like `test_dataloader` is not a kwarg, it's `test_dataloaders` with an 's'. Similar to val_dataloaders, but not the same as train_dataloader without the s, it looks like. Some of the docs are inconsistent on the use of "s" btw, I think.

**Comment by karpathy** [comment]
Ok I think things have improved quite a bit. In particular, my "fake lightning" has now been reduced all the way to 

```python
class LightningModule(nn.Module):
    pass

class Callback:
    pass
```

which is fun :) And I can train with the fake trainer or the lightning trainer and the code looks decent ish.

File: bench.py (hunk 5 above)
Comment (williamFalcon) [reply bench.py]: yes. we enable multiple dataloaders for val and test. coming support for train. 
not in research i’m used to, but turns out some people need two datasets to validate haha. go figure

Comment (karpathy) [reply bench.py]: some people always need something, which is why frameworks are so hard. Next thing you know you can't use a list of data loaders and have to introduce a `DataLoaderSetManager` object.
